Strange borrow check issue

It might be that I'm simply failing to see something blindingly obvious here, but I am having a rather strange issue related to borrow checking. Consider the following very simple function (`nalgebra 0.16.0`): 

```rust
fn modify(y: &mut DVector<f64>) { let slice = y.data.as_mut_slice(); }
```

I get the error:
```
error[E0596]: cannot borrow immutable borrowed content as mutable
 --> src/main.rs:5:47
  |
5 | fn modify(y: &mut DVector<f64>) { let slice = y.data.as_mut_slice(); }
  |                                               ^^^^^^ cannot borrow as mutable

```

I can not for the life of me figure out why this should be problematic. I have a *mutable* reference to a `DVector`, which is essentially just a variant of `Matrix`, which has a field named `data`. I don't see why I cannot borrow this field as mutable. I was unable to reconstruct the problem with standard data structures (`Vec`, slices ...), but I also don't really see how this problem can be specific to `nalgebra`.

I am probably just getting tripped up in some minor detail here. Would appreciate if anyone can point out to me what's going on!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,7 +35,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   * Implement `Extend<Matrix<...>>` for matrices with dynamic storage. This will concatenate the columns of both matrices.
   * Implement `Into<Vec>` for the `MatrixVec` storage.
   * Implement `Hash` for all matrices.
-  
+  * Add a `.len()` method to retrieve the size of a `MatrixVec`.
+
 ### Modified
   * The orthographic projection no longer require that `bottom < top`, that `left < right`, and that `znear < zfar`. The
   only restriction now ith that they must not be equal (in which case the projection would be singular).
@@ -46,11 +47,14 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   * Renamed `.unwrap()` to `.into_inner()` for geometric types that wrap another type.
     This is for the case of `Unit`, `Transform`, `Orthographic3`, `Perspective3`, `Rotation`.
   * Deprecate several functions at the root of the crate (replaced by methods).
-  
+
+### Removed
+    * Remove the `Deref` impl for `MatrixVec` as it could cause hard-to-understand compilation errors.
+
 ### nalgebra-glm
   * Add several alternative projection computations, e.g., `ortho_lh`, `ortho_lh_no`, `perspective_lh`, etc.
   * Add features matching those of nalgebra, in particular: `serde-serialize`, `abmonation-serialize`, std` (enabled by default).
-  
+
 ## [0.16.0]
 All dependencies have been updated to their latest versions.
 

--- a/src/base/construction.rs
+++ b/src/base/construction.rs
@@ -270,7 +270,7 @@ where DefaultAllocator: Allocator<N, R, C>
     /// let vec_ptr = vec.as_ptr();
     ///
     /// let matrix = Matrix::from_vec_generic(Dynamic::new(vec.len()), U1, vec);
-    /// let matrix_storage_ptr = matrix.data.as_ptr();
+    /// let matrix_storage_ptr = matrix.data.as_vec().as_ptr();
     ///
     /// // `matrix` is backed by exactly the same `Vec` as it was constructed from.
     /// assert_eq!(matrix_storage_ptr, vec_ptr);

--- a/src/base/vec_storage.rs
+++ b/src/base/vec_storage.rs
@@ -50,15 +50,16 @@ impl<N, R: Dim, C: Dim> VecStorage<N, R, C> {
 
     /// The underlying data storage.
     #[inline]
-    pub fn data(&self) -> &Vec<N> {
+    pub fn as_vec(&self) -> &Vec<N> {
         &self.data
     }
 
     /// The underlying mutable data storage.
     ///
-    /// This is unsafe because this may cause UB if the vector is modified by the user.
+    /// This is unsafe because this may cause UB if the size of the vector is changed
+    /// by the user.
     #[inline]
-    pub unsafe fn data_mut(&mut self) -> &mut Vec<N> {
+    pub unsafe fn as_vec_mut(&mut self) -> &mut Vec<N> {
         &mut self.data
     }
 

--- a/src/base/vec_storage.rs
+++ b/src/base/vec_storage.rs
@@ -1,6 +1,5 @@
 #[cfg(feature = "abomonation-serialize")]
 use std::io::{Result as IOResult, Write};
-use std::ops::Deref;
 
 #[cfg(all(feature = "alloc", not(feature = "std")))]
 use alloc::vec::Vec;
@@ -81,14 +80,11 @@ impl<N, R: Dim, C: Dim> VecStorage<N, R, C> {
 
         self.data
     }
-}
 
-impl<N, R: Dim, C: Dim> Deref for VecStorage<N, R, C> {
-    type Target = Vec<N>;
-
+    /// The number of elements on the underlying vector.
     #[inline]
-    fn deref(&self) -> &Self::Target {
-        &self.data
+    pub fn len(&self) -> usize {
+        self.data.len()
     }
 }
 
@@ -145,7 +141,7 @@ where DefaultAllocator: Allocator<N, Dynamic, C, Buffer = Self>
 
     #[inline]
     fn as_slice(&self) -> &[N] {
-        &self[..]
+        &self.data
     }
 }
 
@@ -189,7 +185,7 @@ where DefaultAllocator: Allocator<N, R, Dynamic, Buffer = Self>
 
     #[inline]
     fn as_slice(&self) -> &[N] {
-        &self[..]
+        &self.data
     }
 }
 

